File: backend/src/main/java/com/devrats/service/GitHubApiClient.java
package com.devrats.service;

import com.fasterxml.jackson.databind.JsonNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestClient;

@Service
public class GitHubApiClient {
    private static final Logger logger = LoggerFactory.getLogger(GitHubApiClient.class);
    private final RestClient restClient;
    private final String clientId;
    private final String clientSecret;

    public GitHubApiClient(
            @Value("${github.clientId}") String clientId,
            @Value("${github.clientSecret}") String clientSecret,
            RestClient.Builder restClientBuilder) {
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.restClient = restClientBuilder.build();
    }

    public String exchangeCodeForToken(String code, String redirectUri) {
        logger.info("Exchanging OAuth code for token...");
        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
        formData.add("client_id", clientId);
        formData.add("client_secret", clientSecret);
        formData.add("code", code);
        if (redirectUri != null) {
            formData.add("redirect_uri", redirectUri);
        }

        try {
            JsonNode response = restClient.post()
                    .uri("https://github.com/login/oauth/access_token")
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .accept(MediaType.APPLICATION_JSON)
                    .body(formData)
                    .retrieve()
                    .body(JsonNode.class);

            if (response != null && response.has("error")) {
                logger.error("GitHub OAuth error: {} - {}", 
                        response.get("error").asText(), 
                        response.has("error_description") ? response.get("error_description").asText() : "unknown");
                return null;
            }

            return response != null && response.has("access_token") ? response.get("access_token").asText() : null;
        } catch (Exception e) {
            logger.error("Failed to parse GitHub token response", e);
            return null;
        }
    }

    private String getUserEmail(String accessToken) {
        try {
            JsonNode emails = restClient.get()
                    .uri("https://api.github.com/user/emails")
                    .header("Authorization", "Bearer " + accessToken)
                    .header("User-Agent", "DevRats-App")
                    .accept(MediaType.APPLICATION_JSON)
                    .retrieve()
                    .body(JsonNode.class);

            if (emails != null && emails.isArray()) {
                for (JsonNode emailNode : emails) {
                    if (emailNode.has("primary") && emailNode.get("primary").asBoolean()) {
                        return emailNode.get("email").asText();
                    }
                }
                if (!emails.isEmpty()) {
                    return emails.get(0).get("email").asText();
                }
            }
        } catch (Exception e) {
            logger.error("Failed to fetch GitHub emails", e);
        }
        return null;
    }

    public GitHubUser getUserInfo(String accessToken) {
        try {
            JsonNode json = restClient.get()
                    .uri("https://api.github.com/user")
                    .header("Authorization", "Bearer " + accessToken)
                    .header("User-Agent", "DevRats-App")
                    .accept(MediaType.APPLICATION_JSON)
                    .retrieve()
                    .body(JsonNode.class);

            if (json == null || !json.has("id") || !json.has("login")) {
                return null;
            }

            String email = getUserEmail(accessToken);
            if (email == null && json.has("email") && !json.get("email").isNull()) {
                email = json.get("email").asText();
            }

            String name = json.has("name") && !json.get("name").isNull() ? json.get("name").asText() : json.get("login").asText();
            String avatarUrl = json.has("avatar_url") && !json.get("avatar_url").isNull() ? json.get("avatar_url").asText() : null;

            return new GitHubUser(
                    json.get("id").asText(),
                    json.get("login").asText(),
                    name,
                    avatarUrl,
                    email
            );
        } catch (Exception e) {
            logger.error("Failed to fetch GitHub user info", e);
            return null;
        }
    }

    public record GitHubUser(String id, String login, String name, String avatarUrl, String email) {}
}

File: backend/src/main/kotlin/com/devrats/features/github/GitHubApiClient.kt
package com.devrats.features.github

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.request.forms.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.slf4j.LoggerFactory

class GitHubApiClient(private val clientId: String, private val clientSecret: String) {
    private val logger = LoggerFactory.getLogger(GitHubApiClient::class.java)

    private val client = HttpClient(CIO)

    /**
     * Exchange the OAuth code for a GitHub access token.
     * Uses form-urlencoded POST (GitHub's preferred format).
     */
    suspend fun exchangeCodeForToken(code: String): String? {
        logger.info("Exchanging OAuth code for token (clientId=$clientId, code=${code.take(6)}...)")

        val response = client.submitForm(
            url = "https://github.com/login/oauth/access_token",
            formParameters = Parameters.build {
                append("client_id", clientId)
                append("client_secret", clientSecret)
                append("code", code)
            }
        ) {
            accept(ContentType.Application.Json)
        }

        val body = response.bodyAsText()
        val status = response.status
        logger.info("GitHub token exchange response (status=$status): $body")

        return try {
            val json = Json.parseToJsonElement(body).jsonObject

            // Check for error in response
            val error = json["error"]?.jsonPrimitive?.content
            if (error != null) {
                val desc = json["error_description"]?.jsonPrimitive?.content ?: "unknown"
                logger.error("GitHub OAuth error: $error - $desc")
                return null
            }

            json["access_token"]?.jsonPrimitive?.content
        } catch (e: Exception) {
            logger.error("Failed to parse GitHub token response: $body", e)
            null
        }
    }

    /**
     * Fetch the authenticated user's profile from GitHub.
     */
    suspend fun getUserInfo(accessToken: String): GitHubUser? {
        logger.info("Fetching GitHub user info...")

        val response = client.get("https://api.github.com/user") {
            header("Authorization", "Bearer $accessToken")
            header("User-Agent", "DevRats-App")
            accept(ContentType.Application.Json)
        }

        val body = response.bodyAsText()
        logger.info("GitHub user info response (status=${response.status}): ${body.take(200)}")

        return try {
            val json = Json.parseToJsonElement(body).jsonObject
            GitHubUser(
                id = json["id"]?.jsonPrimitive?.content ?: return null,
                login = json["login"]?.jsonPrimitive?.content ?: return null,
                name = json["name"]?.jsonPrimitive?.content ?: json["login"]?.jsonPrimitive?.content ?: "",
                avatarUrl = json["avatar_url"]?.jsonPrimitive?.content
            )
        } catch (e: Exception) {
            logger.error("Failed to parse GitHub user info: $body", e)
            null
        }
    }
}

data class GitHubUser(val id: String, val login: String, val name: String, val avatarUrl: String?)
